File: scripts/test_solutions/127_word_ladder.js
var ladderLength = function(beginWord, endWord, wordList) {
    const wordSet = new Set(wordList);
    if (!wordSet.has(endWord)) return 0;
    const queue = [[beginWord, 1]];
    while (queue.length) {
        const [word, steps] = queue.shift();
        if (word === endWord) return steps;
        for (let i = 0; i < word.length; i++) {
            for (let c = 97; c <= 122; c++) {
                const newWord = word.slice(0, i) + String.fromCharCode(c) + word.slice(i + 1);
                if (wordSet.has(newWord)) {
                    queue.push([newWord, steps + 1]);
                    wordSet.delete(newWord);
                }
            }
        }
    }
    return 0;
};

File: scripts/test_solutions/127_word_ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        if (wordSet.find(endWord) == wordSet.end()) {
            return 0;
        }
        queue<pair<string, int>> q;
        q.push({beginWord, 1});
        while (!q.empty()) {
            auto [word, steps] = q.front();
            q.pop();
            if (word == endWord) {
                return steps;
            }
            for (int i = 0; i < word.size(); i++) {
                string newWord = word;
                for (char c = 'a'; c <= 'z'; c++) {
                    newWord[i] = c;
                    if (wordSet.find(newWord) != wordSet.end()) {
                        q.push({newWord, steps + 1});
                        wordSet.erase(newWord);
                    }
                }
            }
        }
        return 0;
    }
};
